src/clients/inst_autosetup_upgrade.rb

```diff
@@ -295,6 +296,12 @@ def main
         Builtins.y2milestone("Update summary: %1", @update_sum)
         Update.unknown_packages = Ops.get(@update_sum, :ProblemListSze, 0)
 
+        # deselect the upgraded obsolete products (bsc#1133215)
+        Y2Packager::ProductUpgrade.remove_obsolete_upgrades
+        # and select its replacement at first, so later it can be explicitelly removed by user
+        # profile
+        AddOnProduct.missing_upgrades.each { |p| Pkg.ResolvableInstall(p, :product) }
```

Comment:
I think the order should be the opposite, first select the missing upgrades, then remove the obsolete upgrades.

The `missing_upgrades` call does not ignore the add-ons listed in the [`UPGRADE_REMOVAL_MAPPING`](https://github.com/yast/yast-yast2/blob/master/library/packages/src/lib/y2packager/product_upgrade.rb#L52) list so it could possibly select back some add-ons previously deselected by the `ProductUpgrade.remove_obsolete_upgrades` call.

Reply:
hmm, this is quite tricky. I think it should be somewhere documented, but I am not sure where? maybe in both methods yardoc?

Reply:
Um, it's difficult to find the right place as both methods are separate. Ideally we should refactor this to one place and have a single responsibility for this.